sensor/common/processsignal/pipeline_test.go

```diff
@@ -6,17 +6,183 @@ import (
 	"github.com/stackrox/rox/sensor/common/detector/mocks"
 	"github.com/stackrox/rox/sensor/common/message"
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 	"go.uber.org/mock/gomock"
 )
 
-func TestProcessPipelineOffline(t *testing.T) {
+const (
+	containerID1      = "1e43ac4f61f9"
+	containerID2      = "2e43ac4f61f9"
+	containerID3      = "3e43ac4f61f9"
+	deploymentID1     = "mock-deployment-1"
+	deploymentID2     = "mock-deployment-2"
+	deploymentID3     = "mock-deployment-3"
+	outputChannelSize = 2
+)
+
+func TestProcessPipelineOfflineV3(t *testing.T) {
+	t.Setenv(features.SensorCapturesIntermediateEvents.EnvVar(), "true")
+	cases := map[string]struct {
+		entities []clusterentities.ContainerMetadata
+		events   []func(*testing.T, *Pipeline)
+	}{
+		"Online -> signal -> read -> offline -> signal -> online -> read": {
+			entities: []clusterentities.ContainerMetadata{
+				newEntity(containerID1, deploymentID1),
+				newEntity(containerID2, deploymentID2),
+			},
+			events: []func(*testing.T, *Pipeline){
+				online,
+				signal(&storage.ProcessSignal{ContainerId: containerID1}, false),
+				assertSize(1),
+				read(containerID1, deploymentID1),
+				assertSize(0),
+				offline,
+				signal(&storage.ProcessSignal{ContainerId: containerID2}, false),
+				assertSize(1),
+				online,
+				read(containerID2, deploymentID2),
+				assertSize(0),
+			},
+		},
+		"Offline -> signal -> signal -> online -> read -> read": {
+			entities: []clusterentities.ContainerMetadata{
+				newEntity(containerID1, deploymentID1),
+				newEntity(containerID2, deploymentID2),
+			},
+			events: []func(*testing.T, *Pipeline){
+				offline,
+				signal(&storage.ProcessSignal{ContainerId: containerID1}, false),
+				signal(&storage.ProcessSignal{ContainerId: containerID2}, false),
+				assertSize(2),
+				online,
+				read(containerID1, deploymentID1),
+				read(containerID2, deploymentID2),
+				assertSize(0),
+			},
+		},
+		"Offline -> signal -> signal -> signal -> Online -> read -> read": {
+			entities: []clusterentities.ContainerMetadata{
+				newEntity(containerID1, deploymentID1),
+				newEntity(containerID2, deploymentID2),
+				newEntity(containerID3, deploymentID3),
+			},
+			events: []func(*testing.T, *Pipeline){
+				offline,
+				signal(&storage.ProcessSignal{ContainerId: containerID1}, false),
+				signal(&storage.ProcessSignal{ContainerId: containerID2}, false),
+				signal(&storage.ProcessSignal{ContainerId: containerID3}, true),
+				assertSize(2), // The third signal should be dropped
+				online,
+				read(containerID1, deploymentID1),
+				read(containerID2, deploymentID2),
+				assertSize(0), // The buffer should be empty at this point
+			},
+		},
+	}
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			sensorEvents := make(chan *message.ExpiringMessage, outputChannelSize)
+			// defer close(sensorEvents)
```

Comment:
Leftover? Also, rather than using `defer` in tests, you can use `t.Cleanup`

Reply:
I forgot to push the changes to improve the shutdown 🤦 That defer was racing because shutdown does not wait for all the goroutines to finish.